packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -27,138 +28,433 @@ import { parseVatSlot } from './parseVatSlots.js';
+      const parsed = parseVatSlot(vref);
+      assert.equal(parsed.type, 'object', vref);
+      // not allocated by us? that makes it a Presence-type
+      if (!parsed.allocatedByVat) {
+        // if we're dropping the vref, checkImportPresence() already
+        // did our isReachable check, so we can safely skip it (and
+        // save a vatstore syscall)
+        const isReachable =
+          !importsToDrop.has(vref) &&
+          (slotToVal.has(vref) || vrm.isPresenceReachable(vref));
+        // note: the old version used getValForSlot, which treats
+        // COLLECTED as unreachable, which would allow a retirement at
+        // the wrong time (before the finalizer fires and does the
+        // dropImport)
+        const isRecognizable = isReachable || vrm.isVrefRecognizable(vref);
+        if (!isRecognizable) {
+          importsToRetire.add(vref);
         }
       }
-      possiblyRetiredSet.clear();
-
-      const deadBaseRefs = Array.from(deadSet);
-      deadBaseRefs.sort();
-      for (const baseRef of deadBaseRefs) {
-        const { virtual, durable, allocatedByVat, type } =
-          parseVatSlot(baseRef);
-        type === 'object' || Fail`unprepared to track ${type}`;
-        if (virtual || durable) {
-          // Representative: send nothing, but perform refcount checking
-
-          const [gcAgain, retirees] = vrm.deleteVirtualObject(baseRef);
-          if (retirees) {
-            retirees.map(retiree => exportsToRetire.add(retiree));
-          }
-          doMore = doMore || gcAgain;
-        } else if (allocatedByVat) {
-          // Remotable: send retireExport
-          // for remotables, vref === baseRef
-          if (kernelRecognizableRemotables.has(baseRef)) {
-            kernelRecognizableRemotables.delete(baseRef);
-            exportsToRetire.add(baseRef);
-          }
-        } else {
-          // Presence: send dropImport unless reachable by VOM
-          // eslint-disable-next-line no-lonely-if
-          if (!vrm.isPresenceReachable(baseRef)) {
-            importsToDrop.add(baseRef);
-
-            if (!vrm.isVrefRecognizable(baseRef)) {
-              // for presences, baseRef === vref
-              importsToRetire.add(baseRef);
-            }
-          }
-        }
-      }
-    } while (possiblyDeadSet.size > 0 || possiblyRetiredSet.size > 0 || doMore);
+    }
+
+    // note that retiring Presence-type vrefs cannot shake loose any
+    // local data, so we don't need to loop back around
 
     if (importsToDrop.size) {
-      syscall.dropImports(Array.from(importsToDrop).sort());
+      syscall.dropImports([...importsToDrop].sort());
     }
     if (importsToRetire.size) {
-      syscall.retireImports(Array.from(importsToRetire).sort());
+      syscall.retireImports([...importsToRetire].sort());
     }
     if (exportsToRetire.size) {
-      syscall.retireExports(Array.from(exportsToRetire).sort());
+      syscall.retireExports([...exportsToRetire].sort());
     }
```

Comment:
Do we still need to sort these since we sort the inputs above?

Reply:
Hm, maybe not, but it's cheap and safe. We may do multiple rounds of `gcAndFinalize`, but we sort each time, so I guess the most likely source of variability would be if something were shaken loose during round 2 instead of round 3, but that shouldn't happen unless the `gcAndFinalize` is not complete.